compiler/test/dotty/tools/debug/DebugTests.scala

```diff
@@ -0,0 +1,134 @@
+    extends RunTest(testSources, times, threadLimit, suppressAllOutput):
+
+    override def onSuccess(testSource: TestSource, reporters: Seq[TestReporter], logger: LoggedRunnable) =
+      verifyDebug(testSource.outDir, testSource, countWarnings(reporters), reporters, logger)
+
+    private def verifyDebug(dir: JFile, testSource: TestSource, warnings: Int, reporters: Seq[TestReporter], logger: LoggedRunnable) =
+      if Properties.testsNoRun then addNoRunWarning()
+      else
+        val checkFile = testSource.checkFile.getOrElse(throw new Exception("Missing check file")).toPath
+        val debugSteps = DebugStepAssert.parseCheckFile(checkFile)
+        val expressionEvaluator =
+          ExpressionEvaluator(testSource.sourceFiles, testSource.flags, testSource.runClassPath, testSource.outDir)
+        try debugMain(testSource.runClassPath): debuggee =>
+          val jdiPort = debuggee.readJdiPort()
+          val debugger = Debugger(jdiPort, expressionEvaluator, maxDuration/* , verbose = true */)
+          // configure the breakpoints before starting the debuggee
+          val breakpoints = debugSteps.map(_.step).collect { case b: DebugStep.Break => b }.distinct
+          for b <- breakpoints do debugger.configureBreakpoint(b.className, b.line)
+          try
+            debuggee.launch()
+            playDebugSteps(debugger, debugSteps/* , verbose = true */)
+            val status = debuggee.exit()
+            reportDebuggeeStatus(testSource, status)
+          finally
+            // closing the debugger must be done at the very end so that the
+            // 'Listening for transport dt_socket at address: <port>' message is ready to be read
+            // by the next DebugTest
+            debugger.dispose()
+        catch case DebugStepException(message, location) =>
+          echo(s"\n[error] Debug step failed: $location\n" + message)
+          failTestSource(testSource)
+    end verifyDebug
+
+    private def playDebugSteps(debugger: Debugger, steps: Seq[DebugStepAssert[?]], verbose: Boolean = false): Unit =
+      import scala.language.unsafeNulls
+
+      /** The DebugTests can only debug one thread at a time. It cannot handle breakpoints in concurrent threads.
+       *  When thread is null, it means the JVM is running and no thread is waiting to be resumed.
+       *  If thread is not null, it is waiting to be resumed by calling continue, step or next.
+       *  While the thread is paused, it can be used for evaluation.
+       */
+      var thread: ThreadReference = null
+      def location = thread.frame(0).location
+      def continueIfPaused(): Unit =
+        if thread != null then
+          debugger.continue(thread)
+          thread = null
+
+      for case step <- steps do
+        import DebugStep.*
+        try step match
+          case DebugStepAssert(Break(className, line), assert) =>
```

Comment:
Consider
```suggestion
          case DebugStepAssert(Break(className, line), assertion) =>
```
The calls of the form `assert(location)` are confusing because it looks like we're calling `Predef.assert`.

Reply:
done